confluent_kafka_helpers/tracing/backends/opentracer.py

```diff
@@ -0,0 +1,92 @@
+import contextlib
+import io
+import sys
+import traceback
+
+import opentracing
+
+from confluent_kafka_helpers.tracing import tags
+
+DEFAULT_TAGS = {
+    tags.COMPONENT: 'confluent_kafka_helpers',
+    tags.PEER_SERVICE: 'kafka',
+}
+
+
+class OpenTracerBackend:
+    def __init__(self, tracer=opentracing.global_tracer):
+        self._tracer = tracer()
+
+    def __getattr__(self, name):
+        return getattr(self._tracer, name)
+
+    @contextlib.contextmanager
+    def start_active_span(self, *args, **kwargs):
+        # we can't use `start_active_span` or `start_span` built in context
+        # managers since they default catches all exceptions and logs them
+        # incorrectly in `opentracing.span.Span.__exit__` when using the
+        # Datadog tracer.
+        #
+        # https://docs.datadoghq.com/tracing/visualization/trace/?tab=spantags
+        try:
+            scope = self._tracer.start_active_span(
+                *args, tags=DEFAULT_TAGS, **kwargs
+            )
+            yield scope.span
+        finally:
+            self.log_exception(span=scope.span)
+            scope.span.finish()
+
+    @contextlib.contextmanager
+    def start_span(self, *args, **kwargs):
+        try:
+            span = self._tracer.start_span(*args, tags=DEFAULT_TAGS, **kwargs)
+            yield span
+        finally:
+            self.log_exception(span=span)
+            span.finish()
+
+    def extract_headers_and_start_span(self, operation_name, headers):
+        try:
+            parent_context = self._tracer.extract(
+                carrier=headers, format=opentracing.Format.TEXT_MAP
+            )
+        except (
+            opentracing.InvalidCarrierException,
+            opentracing.SpanContextCorruptedException,
+        ):
+            span = self.start_active_span(operation_name=operation_name)
+        else:
+            span = self.start_active_span(
+                operation_name=operation_name,
+                references=[opentracing.follows_from(parent_context)]
+            )
+        return span
+
+    def inject_headers_and_start_span(self, operation_name, headers):
+        span = self.start_span(operation_name=operation_name)
+        self._tracer.inject(
+            span_context=self._tracer.active_span.context, carrier=headers,
+            format=opentracing.Format.TEXT_MAP
+        )
+        return span
+
+    def log_exception(self, span):
+        exc_type, exc_value, exc_tb = sys.exc_info()
+        if not exc_type:
+            return
+
+        buffer = io.StringIO()
+        traceback.print_exception(
+            exc_type, exc_value, exc_tb, file=buffer, limit=20
```

Comment:
What does the limit=20 represent, only print 20 lines of the traceback? Is this something that would make sense to make configurable? Or is 20 limit good enough for most use cases?

Reply:
It's the number of stack traces not lines. So for example this would generate 2 stack traces:
```python
try:
   raise Foo
except Exception as e:
   raise Bar(e)
```
It seems like Datadog also parses the stack traces so it never displays more then 5 lines for each stack trace + one extra line for the last raised exception. Here is an example for two stack traces:

![image](https://user-images.githubusercontent.com/853582/101783835-07620a00-3afb-11eb-9e33-0ba575d08cc6.png)

I don't think this have to be configurable at the moment, I think 20 stack traces are more then enough to understand what the problem is. What do you think?

Reply:
Ah got it, yea 20 sounds like good enought.